File: leetcode/1329-sort-the-matrix-diagonally/solution.py
from typing import List
from collections import defaultdict


class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        m, n = len(mat), len(mat[0])
        diagonals = defaultdict(list)

        for i in range(m):
            for j in range(n):
                diagonals[i - j].append(mat[i][j])

        for key in diagonals:
            diagonals[key].sort(reverse=True)  # pop() from the end gives ascending order

        for i in range(m):
            for j in range(n):
                mat[i][j] = diagonals[i - j].pop()

        return mat

File: leetcode/1329-sort-the-matrix-diagonally/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        unordered_map<int, vector<int>> diagonals;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                diagonals[i - j].push_back(mat[i][j]);
            }
        }

        for (auto& [key, vals] : diagonals) {
            sort(vals.rbegin(), vals.rend());
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                mat[i][j] = diagonals[i - j].back();
                diagonals[i - j].pop_back();
            }
        }

        return mat;
    }
};
